firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalSerializer.java

```diff
@@ -266,4 +269,32 @@ TargetData decodeTargetData(com.google.firebase.firestore.proto.Target targetPro
         lastLimboFreeSnapshotVersion,
         resumeToken);
   }
+
+  public com.google.firestore.proto.BundledQuery encodeQuery(Query query) {
+    // We store LimitToLast as metadata information below. We remove it from the query itself as
+    // `toTarget()` flips the order by constraints otherwise.
+    Query queryWithoutLimitToLast =
```

Comment:
I think we could avoid this by parsing json into proto class directly. This is required here because `Query` the model object has its own handling logic for limit to last. 

In order to make the interface of `BundleSerializer` consistent, we can introduce a new `local.BundledQuery` which holds a proto `StructuredQuery`. `local.NamedQuery` holds a `BundledQuery` instead of `Query`.

In `BundleSerializer` the interface stays unchanged, but the implementation will work with `StructuredQuery` directly, without having to deal with `Query`.

This way, the `encodeQuery` here can be eliminated. We can also have the test for limit to last queries the same way as in JS, constructing a `NamedQuery.bundledQuery` directly and save it, read it back from persistence and make sure it is a proper limit to last query.

Reply:
There are two problems I encountered:

- It it is not possible to instantiate a Query object from a Target. If that was possible, `BundledQuery` would contain a Target with an additional bit that describes the limit. 
- It is also not possible to get the original order constraints from the query, since `toTarget()` flips the ordering constraint. The hack here aims to undo this.

Furthermore, the Android SDK cannot deal with StructuredQuery directly. Instead, it expects the data types that we use in Target. If we don't deserialize here, we have to at another point (e.g. to allow a user to modify a named query).

To clean this up, I think we need to somehow figure out how to change Query to only contain a Target and a Limit flag. The target object would then contain the constraints and filters. This would allow us to construct a Query from a Target and to retrieve the Target from a Query without modifying it.

Anyways, I will schedule some time to discuss.

Reply:
I updated this based on our offline discussion and replaces the `queryWithoutLimitToLast` without something a little less hacky. Hope this works :)